providers/implementations/keymgmt/ml_kem_kmgmt.c

```diff
@@ -0,0 +1,631 @@
+    /*
+     * If the private key is given, we'll ignore the public key data, taking
+     * the embedded public key as authoritative.
+     */
+    if (prvlen != 0) {
+        if (ossl_ml_kem_vparse_private_key(v, &key->prvkey, prvenc, prvlen,
+                                           key->ctx) == 1)
+            return 1;
+        ossl_ml_kem_vcleanse_prvkey(v, &key->prvkey);
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY);
+        return 0;
+    }
+
+    if (ossl_ml_kem_vparse_public_key(v, &key->pubkey, pubenc, publen,
+                                      key->ctx) == 1)
+        return 1;
+    OPENSSL_free(key->pubkey);
+    key->pubkey = NULL;
+    ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY);
+    return 0;
+}
+
+static int ml_kem_import(void *vkey, int selection, const OSSL_PARAM params[])
+{
+    ML_KEM_PROVIDER_KEYPAIR *key = vkey;
+    int include_private;
+
+    if (!ossl_prov_is_running() || key == NULL)
+        return 0;
+
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
+        return 0;
+
+    include_private = selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY ? 1 : 0;
+    return ossl_ml_kem_key_fromdata(key, params, include_private);
+}
+
+static const OSSL_PARAM *ml_kem_gettable_params(void *provctx)
+{
+    static const OSSL_PARAM arr[] = {
+        OSSL_PARAM_int(OSSL_PKEY_PARAM_BITS, NULL),
+        OSSL_PARAM_int(OSSL_PKEY_PARAM_SECURITY_BITS, NULL),
+        OSSL_PARAM_int(OSSL_PKEY_PARAM_MAX_SIZE, NULL),
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY, NULL, 0),
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_PRIV_KEY, NULL, 0),
+        OSSL_PARAM_END
+    };
+
+    return arr;
+}
+
+/*
+ * It is assumed the key is guaranteed non-NULL here, and is from this provider
+ */
+static int ml_kem_get_params(void *vkey, OSSL_PARAM params[])
+{
+    ML_KEM_PROVIDER_KEYPAIR *key = vkey;
+    vinfo_t v = key->vinfo;
+    OSSL_PARAM *p;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_BITS);
+    if (p != NULL)
+        if (!OSSL_PARAM_set_int(p, v->bits))
+            return 0;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_SECURITY_BITS);
+    if (p != NULL)
+        if (!OSSL_PARAM_set_int(p, v->secbits))
+            return 0;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_MAX_SIZE);
+    if (p != NULL)
+        if (!OSSL_PARAM_set_int(p, v->ctext_bytes))
+            return 0;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY);
+    if (p != NULL && have_keys(key)) {
+        if (p->data_type != OSSL_PARAM_OCTET_STRING)
+            return 0;
+        p->return_size = key->vinfo->pubkey_bytes;
+        if (p->data != NULL) {
+            if (p->data_size < p->return_size)
+                return 0;
+            if (!ossl_ml_kem_vencode_public_key(v, p->data, p->return_size,
+                                                key->pubkey, key->prvkey))
+                return 0;
+        }
+    }
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_PRIV_KEY);
```

Comment:
@mattcaswell Should we have a OSSL_PKEY_PARAM_ENCODED_PRIVATE_KEY also?

Reply:
May I suggest to take this discussion out of this PR and align with/implement only when closing https://github.com/openssl/openssl/issues/25885 (discussion on which private key format to "externalize": Basically, NIST or IETF)?

Reply:
> @mattcaswell Should we have a OSSL_PKEY_PARAM_ENCODED_PRIVATE_KEY also?

We don't have that anywhere else - but lots of instances of `OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY`. The idea behind the public key version is to easily enable the sharing of the public key between peers in some protocol. Originally this was TLS specific (in 1.1.1 we had `EVP_PKEY_get1_tls_encodedpoint`), but at some point we made the naming more generic. Since you never share the private key in that way there has never been a need for `OSSL_PKEY_PARAM_ENCODED_PRIVATE_KEY`.